src/jmh/java/dev/blaauwendraad/masker/json/BenchmarkUtils.java

```diff
@@ -64,9 +90,72 @@ public static String randomJson(Set<String> targetKeys, String jsonSize, String
+        collectJsonPaths(jsonNode, jsonPaths, "$", lowerCaseTargetKeys);
+        return jsonNode;
+    }
+
+    private static void collectJsonPaths(JsonNode jsonNode, Set<String> jsonPaths, String currentJsonPath, Set<String> targetKeys) {
+        if (jsonNode instanceof ObjectNode objectNode) {
+            Iterable<String> fieldNames = objectNode::fieldNames;
+            for (String fieldName : fieldNames) {
+                if (fieldName.isEmpty() || fieldName.startsWith(".") || fieldName.endsWith(".") || fieldName.contains("..")) {
+                    // such filed would result in descendant segment '..' that we don't support
+                    continue;
+                }
+                String jsonPathKey = currentJsonPath + "." + fieldName;
+                if (targetKeys.contains(fieldName.toLowerCase())) {
+                    jsonPaths.add(jsonPathKey);
+                }
+                collectJsonPaths(jsonNode.get(fieldName), jsonPaths, jsonPathKey, targetKeys);
+            }
+        } else if (jsonNode instanceof ArrayNode arrayNode) {
+            String jsonPathKey = currentJsonPath + "[*]";
+            for (int i = 0; i < arrayNode.size(); i++) {
+                collectJsonPaths(arrayNode.get(i), jsonPaths, jsonPathKey, targetKeys);
+            }
+        }
     }
+
+    public static Set<String> collectJsonPaths(JsonNode jsonNode, Set<String> targetKeys) {
+        Set<String> jsonPaths = new HashSet<>();
+        Set<String> lowerCaseTargetKeys = targetKeys.stream().map(String::toLowerCase).collect(Collectors.toSet());
+        collectJsonPaths(jsonNode, jsonPaths, "$", lowerCaseTargetKeys);
+        return disambiguate(jsonPaths);
+    }
+
+    /**
+     * Disambiguates the input list of jsonpath keys.
+     * <p>
+     * If two keys <code>$.a.b</code> and <code>$.a</code> are present, the more specific <code>$.a.b</code> will be chosen.
+     *
+     * @param jsonPaths the input list of jsonpath keys
+     * @return disambiguated set of jsonpath keys
+     */
+public static Set<String> disambiguate(Set<String> jsonPaths) {
+    while (true) {
+        Set<String> disambiguated = new HashSet<>(jsonPaths);
+        List<String> jsonPathKeys = new ArrayList<>(jsonPaths);
+        Collections.sort(jsonPathKeys);
+        JsonPathParser jsonPathParser = new JsonPathParser();
+        for (int i = 1; i < jsonPathKeys.size(); i++) {
+            String current = jsonPathKeys.get(i - 1);
+            String next = jsonPathKeys.get(i);
+            try {
+                jsonPathParser.checkAmbiguity(Set.of(jsonPathParser.parse(current), jsonPathParser.parse(next)));
+            } catch (IllegalArgumentException ignore) {
+                disambiguated.remove(current);
+            }
+        }
+        if (disambiguated.size() == jsonPaths.size()) {
+            return disambiguated;
+        }
+        jsonPaths = disambiguated;
+    }
+}
```

Comment:
```suggestion
    public static Set<String> disambiguate(Set<String> jsonPaths) {
        while (true) {
            Set<String> disambiguated = new HashSet<>(jsonPaths);
            List<String> jsonPathKeys = new ArrayList<>(jsonPaths);
            Collections.sort(jsonPathKeys);
            JsonPathParser jsonPathParser = new JsonPathParser();
            for (int i = 1; i < jsonPathKeys.size(); i++) {
                String current = jsonPathKeys.get(i - 1);
                String next = jsonPathKeys.get(i);
                try {
                    jsonPathParser.checkAmbiguity(Set.of(jsonPathParser.parse(current), jsonPathParser.parse(next)));
                } catch (IllegalArgumentException ignore) {
                    disambiguated.remove(current);
                }
            }
            if (disambiguated.size() == jsonPaths.size()) {
                return disambiguated;
            }
            jsonPaths = disambiguated;
        }
    }
```

Reply:
Just formatting. I tried to break that `while(true)` loop, but the problem turned out to be more elaborate. If we have [A, B, C, D] json paths and A, B is not ambiguate and B,C is not ambiguate but C,D is ambiguate, we need to check again if B,D and A,D are ambiguate. 
We could somehow sort json paths by their "ambiguity potential", but that's very nuanced and while(true) loop works fine anyway.